zerver/views/streams.py

```diff
@@ -607,16 +616,39 @@ def get_streams_backend(
     return json_success({"streams": streams})
 
 @has_request_variables
-def get_topics_backend(request: HttpRequest, user_profile: UserProfile,
-                       stream_id: int=REQ(converter=to_non_negative_int,
-                                          path_only=True)) -> HttpResponse:
-    (stream, recipient, sub) = access_stream_by_id(user_profile, stream_id)
+def get_topics_backend(
+        request: HttpRequest, maybe_user_profile: Union[UserProfile, AnonymousUser],
+        stream_id: int=REQ(converter=to_non_negative_int, path_only=True),
+        narrow: Optional[List[Dict[str, Any]]]=REQ('narrow', converter=narrow_parameter, default=None),
+) -> HttpResponse:
 
-    result = get_topic_history_for_stream(
-        user_profile=user_profile,
-        recipient=recipient,
-        public_history=stream.is_history_public_to_subscribers(),
-    )
+    if is_web_public_narrow(narrow) and not maybe_user_profile.is_authenticated:
+        is_web_public_query = True
+        user_profile: Optional[UserProfile] = None
+    else:
+        is_web_public_query = False
+        assert isinstance(maybe_user_profile, UserProfile)
+        user_profile = maybe_user_profile
+        assert user_profile is not None
```

Comment:
So in particular we can just have the conditionals be `maybe_user_profile.is_authenticated`